### PR TITLE
Validate the use of style props during reparenting.

### PR DESCRIPTION
**Pull request by seanparsons**
**Problem:**
If a component does not carry positional properties (like `left`) from the `style` attribute into the root element making that component up, then it very likely will not be repositionable. We want something that can't be repositioned to also not be reparentable as well.

**Fix:**
Some very basic static analysis is performed to see if the appropriate properties are carried into the root element's style attribute or if the `style` attribute of the component props are spread into the `style` attribute of the root element.

**Commit Details:**
- Implemented `componentHonoursPropsPosition` and `componentHonoursPropsSize`
  which check if a component honours the position and size properties
  respectively by passing them into the root element.
- Created `propsStyleIsSpreadInto` which checks if the attributes provided
  have `props.style` spread into them.
- Added `propertyComesFromPropsStyle` which analyses the attributes of
  an element to see if a certain property has been utilised from the props
  parameter.
- Added `targetHonoursPropsSize` and `targetHonoursPropsPosition` to
  `MetadataUtils` which provide a higher level call over the
  `componentHonours...` functions.
- Added `honoursPropsPosition` and `honoursPropsSize` to simplify
  checking for canvas strategies.
- Prevent component instances from being reparented if they do not honour
  the position properties in the root element of the component.

### CODE REVIEW

File: editor/src/core/model/element-template-utils.spec.ts [hunk 1]
```diff
@@ -362,3 +364,117 @@ export const TestComponent = (props) => {
...
   })
 })
+
+describe('componentHonoursPropsPosition', () => {
+  it('returns true for a component that uses top and left directly', () => {
+    const component = getComponentFromCode(
+      'TestComponent',
+      `
+export const TestComponent = (props) => {
+  return <div style={{position: 'absolute', left: props.style.left, top: props.style.top}}>The Best Test Component</div>
```
Comment (balazsbajorics) [inline editor/src/core/model/element-template-utils.spec.ts:374]: aww The Best Test Component

File: editor/src/core/model/element-template-utils.spec.ts [hunk 2]
```diff
@@ -362,3 +364,117 @@ export const TestComponent = (props) => {
     expect(result).toEqual(true)
   })
 })
+
+describe('componentHonoursPropsPosition', () => {
```
Comment (balazsbajorics) [inline editor/src/core/model/element-template-utils.spec.ts:368]: could you insert a few test cases for false results? just to make sure the function doesn't accidentally regress into "always true"

**Review by balazsbajorics — APPROVED** [review]
nice! (just in case you don't plan it already) I think a follow-up PR should also disable moving and resizing strategies for elements that don't respect the relevant props

**Comment by seanparsons** [comment]
> nice! (just in case you don't plan it already) I think a follow-up PR should also disable moving and resizing strategies for elements that don't respect the relevant props. IIRC the Delta-based absolute move doesn't need the size props, and there used to be a delta-based absolute resize but maybe that has been removed.

That seems like a natural progression, I thought I would keep it out of this PR since that's not the purpose of it.

File: editor/src/core/model/element-template-utils.spec.ts (hunk 2 above)
Comment (seanparsons) [reply editor/src/core/model/element-template-utils.spec.ts:368]: Additional test cases added in 6744435ba.

**Pull request merged**